src/circuits/mimc/mimc.tcc

```diff
@@ -0,0 +1,167 @@
+// DISCLAIMER:
+// Content taken and adapted from:
+// https://github.com/HarryR/ethsnarks/blob/master/src/gadgets/mimc.hpp
+
+#ifndef __ZETH_MIMC_PERMUTATION_TCC__
+#define __ZETH_MIMC_PERMUTATION_TCC__
+
+namespace libzeth {
+    template<typename FieldT>
+    void MiMCe7_permutation_gadget<FieldT>::_setup_gadgets(
+        const libsnark::pb_variable<FieldT> in_x,
+        const libsnark::pb_variable<FieldT> in_k)
+    {
+        for( size_t i = 0; i < ROUNDS; i++ )
+        {
+            // setting the input of the next round with the output variable of the previous round (except for round 0)
+            const auto& round_x = (i == 0 ? in_x : round_gadgets.back().result() );
```

Comment:
I think it'd be good to avoid to use `auto` in such cases. I personally tend to like being verbose with types as it makes it easier (at least for me) to follow the code and understand it + catch weird scenarios and mistakes. However, using `auto` in for loops like in:
```
      for( auto& gadget : round_gadgets )
        {
            gadget.generate_r1cs_witness();
        }
```
Is quite clear

Reply:
Yeah I see what you mean: we can definitely avoid to use it. If in this special case is ok i leave it as it is.